### ISSUE TITLE
Inherit `ScopeFlags::StrictMode` from parent scope

### ISSUE THREAD

**Issue by overlookmotel** [oxc-project/oxc#3862]
There appear to be some discrepancies in when `ScopeFlags::StrictMode` is applied by semantic.

`enter_scope` *looks like* its logic is that `StrictMode` should be set on any scope which is strict mode (including from a `'use strict'` directive somewhere above it). But this doesn't appear to be what it actually does, because it only sets `StrictMode` flag if the *direct* parent is a function with a `'use strict'` directive, or if the entire program is strict mode:

https://github.com/oxc-project/oxc/blob/9e148e97a95f55d49e5976524242d9b243e0c72c/crates/oxc_semantic/src/builder.rs#L414-L440

A couple of examples of weirdness:

```js
// Script - sloppy mode
function foo() {
  'use strict';
  {
    // Scope flags for this block include `StrictMode`
    {
      // Scope flags for this block do NOT include `StrictMode`
    }
  }
}
```

```js
// Script - sloppy mode
class X {
  foo() {
    // Scope flags for this method do NOT include `StrictMode`
  }
}
```

When is `StrictMode` flag *meant* to be set?

If intended behavior is that `StrictMode` flag should be set for any scope which runs in strict context, we can simplify the above logic to:

```rs
if let Some(parent_scope_id) = parent_scope_id {
    let parent_scope_flags = self.scope.get_flags(parent_scope_id);
    flags |= parent_scope_flags & ScopeFlags::StrictMode;

    // inherit flags for non-function scopes
    if !flags.contains(ScopeFlags::Function) {
        flags |= parent_scope_flags & ScopeFlags::Modifiers;
    }
}
```

**Comment by overlookmotel** [comment]
@overlookmotel

**Comment by overlookmotel** [comment]
NB: #3908 moves the logic above to [oxc_semantic/src/scope.rs](https://github.com/oxc-project/oxc/blob/e4b4cdf3bc7b311766af54730dda3940133361f6/crates/oxc_semantic/src/scope.rs#L102-L126)

**Comment by Boshen** [comment]
strict mode should be inherited, seems like a bug in places that are not checked.


This can be verified by 

https://github.com/oxc-project/oxc/blob/c26975ab7610f65eece293a91481e2a8bd5f517e/crates/oxc_semantic/src/checker/javascript.rs#L146-L149

where strictness is checked on the scope.

**Comment by overlookmotel** [comment]
Thanks for clarifying. I think we can fix with:

1. the replacement mentioned above.
2. set `StrictMode` flag on all classes (NB `extends` clause runs in strict mode, as well as the class body, but I'm not sure about decorators).

**Comment by Boshen** [comment]
Seems fixed?

**Comment by overlookmotel** [comment]
Fixed in #4664.
